### PR TITLE
#515 DigestOf (MD5, SHA-1 and SHA-256)

### PR DESCRIPTION
**Pull request by fabriciofx**
As per #515 .

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#516` is now in scope.

File: src/main/java/org/cactoos/io/DigestOf.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    public DigestOf(final Input input) {
+        this(input, "SHA-256");
+    }
+
+    /**
+     * Ctor.
+     * @param input The input
+     * @param algorithm The algorithm
+     */
+    public DigestOf(final Input input, final String algorithm) {
```
Comment (yegor256) [inline src/main/java/org/cactoos/io/DigestOf.java]: @fabriciofx the name `algorithm` is an attribute and an argument. Would be better to use different names.

**Comment by yegor256** [comment]
@fabriciofx looks very good, there is one small comment. Also, I would suggest to write a test for an empty input: `new DigestOf(new InputOf(""))`

File: src/main/java/org/cactoos/io/DigestOf.java (hunk 1 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/io/DigestOf.java]: @yegor256 I'll fix it now. Thanks!

**Comment by fabriciofx** [comment]
@yegor256 Ok. I'll add this test too. Thanks!

File: src/main/java/org/cactoos/io/DigestOf.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+     * The algorithm.
+     */
+    private final String algorithm;
+
+    /**
+     * Ctor.
+     * @param input The input
+     */
+    public DigestOf(final Input input) {
+        this(input, "SHA-256");
```
Comment (Timmeey) [inline src/main/java/org/cactoos/io/DigestOf.java]: using plain strings to represent/select settings was always a major pain for me (was ist "sha-256" or "SHA-256" or was ist "sha256"?. I understand why it can be beneficial to allow plain strings, but could we also provide a Ctor which takes an enum? This makes it easier to select a algorithm. By adding a Ctor which takes an enum we do not prevent ppl to use plain strings if they want.

Provide a few ENUM representing commonly used and available algorithms like MD5/SHA256

Comment (fabriciofx) [reply src/main/java/org/cactoos/io/DigestOf.java]: @Timmeey Yes, there is this problem, but if I'm not wrong, Yegor don't like `enums`. @yegor256 WDYT? Should I put enums in this code? (Maybe a post about why enums are a mistake would be nice.)

Comment (Timmeey) [reply src/main/java/org/cactoos/io/DigestOf.java]: Enums are just an easy way of showing the developer which values are available. In my opinion it is WAY worse to force the developer to look up possible values by going through the sourcecode.

If we don't want enums, we need classes like `SHA256DigestOf` `MD5DigestOf` and then one where you have to provie the algorithm name. THis could then be done by having an abstract DigestOf class with the logic inside, and concrete classes that only provide their appropriate Algorithm string to the abstract DigestOf class. It's either specific classes or enums.

Plain strings as the only way to select the algorithm are just a pain.

Comment (fabriciofx) [reply src/main/java/org/cactoos/io/DigestOf.java]: @Timmeey I agree with this solution. I think it better than enums. I'll provide them. Thanks!

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#516` assigned to [@icok](http://www.0crat.com/u/icok}) (`DEV`). The budget is [fixed](http://datum.zerocracy.com/pages/policy.html#4) and it is 30 minutes. Please, read [§4](http://datum.zerocracy.com/pages/policy.html#4), [§8](http://datum.zerocracy.com/pages/policy.html#8), and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by iCok** [comment]
@0crat waiting

**Comment by iCok** [comment]
@0crat waiting

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-356527114))

@iCok The impediment was registered successfully

**Comment by 0crat** [comment]
> @0crat waiting ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-357006822))

@iCok I can't do it for technical reasons, I'm very sorry.  If you don't know what to do,  submit this error as a ticket  [here](https://github.com/zerocracy/datum):

```
 1.0-SNAPSHOT 9415ec2a 2018-01-11T17:33:30
IllegalStateException: java.lang.reflect.InvocationTargetException in com/zerocracy/stk/pm/in/impediments/register_impediment.groovy
java.lang.IllegalStateException: java.lang.reflect.InvocationTargetException in com/zerocracy/stk/pm/in/impediments/register_impediment.groovy
	at com.zerocracy.farm.reactive.StkGroovy.process(StkGroovy.java:125)
	at com.zerocracy.farm.StkSafe.process(StkSafe.java:89)
	at com.zerocracy.farm.reactive.StkSmart.process(StkSmart.java:66)
	at com.zerocracy.farm.reactive.Brigade.process(Brigade.java:82)
	at com.zerocracy.farm.reactive.Brigade.lambda$apply$0(Brigade.java:64)
	at org.cactoos.func.UncheckedFunc.lambda$apply$0(UncheckedFunc.java:58)
	at org.cactoos.func.IoCheckedBiFunc.apply(IoCheckedBiFunc.java:66)
	at org.cactoos.func.UncheckedBiFunc.apply(UncheckedBiFunc.java:60)
	at org.cactoos.func.UncheckedFunc.apply(UncheckedFunc.java:59)
	at org.cactoos.iterator.Filtered.hasNext(Filtered.java:90)
	at org.cactoos.iterator.LengthOf.lambda$new$0(LengthOf.java:52)
	at org.cactoos.scalar.UncheckedScalar.lambda$value$0(UncheckedScalar.java:58)
	at org.cactoos.func.UncheckedFunc.lambda$apply... 
```

Cc [@yegor256](http://www.0crat.com/u/yegor256)

**Comment by fabriciofx** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-357194940))

@fabriciofx This is what I know about this job, as in [§32](http://datum.zerocracy.com/pages/policy.html#32):

  * The job [#516](https://github.com/yegor256/cactoos/issues/516) is in scope
  * The job is assigned to [@icok](http://www.0crat.com/u/icok)
  * These users are banned: [@fabriciofx](http://www.0crat.com/u/fabriciofx): This user reported the ticket
  * The job has an impediment

**Comment by 0crat** [comment]
[@icok](http://www.0crat.com/u/icok) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://datum.zerocracy.com/pages/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by iCok** [comment]
@yegor256 requirements for that task are not clear, need help.

**Comment by yegor256** [comment]
@iCok just do the code review, what's not clear?

**Comment by iCok** [comment]
@yegor256 Everything good, until I do not have a Rev role. That is why, it is not obvious for me to make a review.

**Comment by yegor256** [comment]
@iCok got it. Just refuse it.

**Comment by iCok** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-357713767))

@iCok [@icok](http://www.0crat.com/u/icok) resigned from [#516](https://github.com/yegor256/cactoos/issues/516), please stop working

**Comment by 0crat** [comment]
The job [#516](https://github.com/yegor256/cactoos/issues/516) assigned to @allanws[/z](http://www.0crat.com/u/allanws). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

File: src/main/java/org/cactoos/io/DigestEnvelope.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+     * The algorithm.
+     */
+    private final String algorithm;
+
+    /**
+     * Ctor.
+     * @param input The input
+     * @param algo The algorithm
+     */
+    public DigestEnvelope(final Input input, final String algo) {
```
Comment (allanws) [inline src/main/java/org/cactoos/io/DigestEnvelope.java]: maybe renaming to "algrthm" (rather than "algo") would be more meaningful? (just a suggestion)

File: src/main/java/org/cactoos/io/DigestEnvelope.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        this(input, 16 << 10, algo);
+    }
+
+    /**
+     * Ctor.
+     * @param input The input
+     * @param max Buffer size
+     * @param algo The algorithm
+     */
+    public DigestEnvelope(final Input input, final int max, final String algo) {
```
Comment (allanws) [inline src/main/java/org/cactoos/io/DigestEnvelope.java]: maybe renaming to "algrthm" (rather than "algo") would be more meaningful? (just a suggestion)

**Comment by allanws** [comment]
@fabriciofx please see my comments above. Also, the build is failing, please fix it.

File: src/main/java/org/cactoos/io/DigestEnvelope.java (hunk 3 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/io/DigestEnvelope.java]: @allanws I agree.

File: src/main/java/org/cactoos/io/DigestEnvelope.java (hunk 4 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/io/DigestEnvelope.java]: @allanws I agree.

**Comment by fabriciofx** [comment]
@allanws I think these kind of tests
```
new Md5DigestOf(null);
new Sha1DigestOf(null)
```
not make sense, because if must check no null inputs, I should use something like:
```
new Md5DigestOf(new Input.NoNulls(new InputOf(...)))
```

File: src/main/java/org/cactoos/io/DigestEnvelope.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        this.source = input;
+        this.size = max;
+        this.algorithm = algrthm;
+    }
+
+    @Override
+    public byte[] asBytes() throws IOException {
+        try {
+            final MessageDigest msg = MessageDigest.getInstance(this.algorithm);
+            try (final InputStream stream = this.source.stream()) {
```
Comment (allanws) [inline src/main/java/org/cactoos/io/DigestEnvelope.java]: so, if you don't add a check for null, this will raise a NPE. Is it OK?

File: src/test/java/org/cactoos/io/Md5DigestOfTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Test
+    public void checksumFromFile() throws IOException {
+        MatcherAssert.assertThat(
+            "Can't calculate the file's MD5 checksum",
+            new HexOf(
+                new Md5DigestOf(
+                    new InputOf(
+                        new ResourceOf(
+                            "org/cactoos/io/DigestEnvelope.class"
```
Comment (allanws) [inline src/test/java/org/cactoos/io/Md5DigestOfTest.java]: I suggest changing this classe for a core java one (this DigestEnvelop.class would be more prone to be removed rather than java.lang.Object)

File: src/main/java/org/cactoos/io/DigestEnvelope.java (hunk 5 above)
Comment (fabriciofx) [reply src/main/java/org/cactoos/io/DigestEnvelope.java]: @allanws Following @yegor256, yes, it's right. If you want to check the input will should use `Input.NoNulls`.

File: src/test/java/org/cactoos/io/Md5DigestOfTest.java (hunk 6 above)
Comment (fabriciofx) [reply src/test/java/org/cactoos/io/Md5DigestOfTest.java]: @allanws If `DigestEnvelope.class` will removed, these tests will be too. :)

**Comment by yegor256** [comment]
@0crat assign @amihaiemil

**Comment by amihaiemil** [comment]
@yegor256 This is reviewd as well, the CR asked for your approval; why assign me, I'm not an architect? seems like double work :)

**Comment by 0crat** [comment]
> @0crat assign @amihaiemil ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-358948404))

@yegor256 The job [#516](https://github.com/yegor256/cactoos/issues/516) assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://datum.zerocracy.com/pages/policy.html#19): -5 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+7685](https://www.0crat.com/u/yegor256)

**Comment by amihaiemil** [comment]
@allanws did you stop working with Zerocracy? should I do another review here?

**Comment by amihaiemil** [comment]
@0crat resign

**Comment by 0crat** [comment]
> @0crat resign ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-360251659))

@amihaiemil The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#516](https://github.com/yegor256/cactoos/issues/516), please stop working

**Comment by 0crat** [comment]
Tasks refusal is discouraged, see [§6](http://datum.zerocracy.com/pages/policy.html#6): -15 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+75](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
The job [#516](https://github.com/yegor256/cactoos/issues/516) assigned to @rexim[/z](https://www.0crat.com/u/rexim). The budget is 30 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§8](http://datum.zerocracy.com/pages/policy.html#8) and [§9](http://datum.zerocracy.com/pages/policy.html#9). If the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html).

**Comment by rexim** [comment]
@yegor256 I don't have a REV role. I believe I was assigned by mistake because of #462. Refusing.

**Comment by rexim** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/cactoos/issues/516#issuecomment-360435864))

@rexim The user @rexim[/z](https://www.0crat.com/u/rexim) resigned from [#516](https://github.com/yegor256/cactoos/issues/516), please stop working

**Comment by 0crat** [comment]
Tasks refusal is discouraged, see [§6](http://datum.zerocracy.com/pages/policy.html#6): -15 points just awarded to @rexim[/z](https://www.0crat.com/u/rexim), total is [-15](https://www.0crat.com/u/rexim)

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#516` is not assigned, can't get start time

**Comment by 0crat** [comment]
The job [#516](https://github.com/yegor256/cactoos/issues/516) is now out of scope

**Comment by 0crat** [comment]
Job [#516](https://github.com/yegor256/cactoos/issues/516) was not in scope
